[git-webkit] Support pushing commits to user-specified remote

JonWBedard opened the pull request:
#### a10d2e295345449f0213e2141054377d23a6c41f
<pre>
[git-webkit] Support pushing commits to user-specified remote
<a href="https://bugs.webkit.org/show_bug.cgi?id=284149">https://bugs.webkit.org/show_bug.cgi?id=284149</a>
<a href="https://rdar.apple.com/141025958">rdar://141025958</a>

Reviewed by Sam Sneddon.

* Tools/Scripts/hooks/pre-push: Allow pushing to unregistered remotes, so long as we treat
those remotes as public.

Canonical link: <a href="https://commits.webkit.org/287491@main">https://commits.webkit.org/287491@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b6a9fe88b3ecbdefe8a6778ede9aca3a7735ff52

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/79891 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/58889 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/33289 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/84406 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/30870 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/82000 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/67953 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/7181 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/62434 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/30870 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/82959 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/52488 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/72750 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/42745 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/49832 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/26899 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/29328 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/70953 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/27376 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/85838 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/7109 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/4981 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/70707 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/7286 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/68590 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/69949 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/12872 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12350 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/7073 "Built successfully") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/6923 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/10434 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/8727 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

gsnedders commented on Tools/Scripts/hooks/pre-push:
```diff
@@ -294,8 +294,9 @@ def main(name, remote):
             return 255
         name = remote_to_name.get('{}:{}'.format(match.group('host'), match.group('path')), None)
         if not name:
+            name = 'user-specified'
```
Is there any reason why we need this? This doesn't appear to some special known-value that appears elsewhere, can we not just keep using the specified URL?

JonWBedard replied:
No, that's a good point. We can use the remote. Will update.

gsnedders replied:
(Pedantic terminology point: `git-push(1)` claims `<repository>` is either a URL or a remote, and we're likely in the former case at this point.)

gsnedders commented on Tools/Scripts/hooks/pre-push:
```diff
@@ -292,11 +292,9 @@ def main(name, remote):
-        name = remote_to_name.get('{}:{}'.format(match.group('host'), match.group('path')), None)
-        if not name:
+        name = remote_to_name.get('{}:{}'.format(match.group('host'), match.group('path')), remote)
+        if name not in remote_to_name.values():
```
```suggestion
        if name not in name_to_remote:
```

No point in doing the relatively slow thing here. But r+'d if you fix that.

JonWBedard replied:
We need to search `values()` not `keys()`, though. I could change this to:
```
name = remote_to_name.get('{}:{}'.format(match.group('host'), match.group('path')), None)
if not name:
    name = remote
    ...
```
is that more clear maybe?

gsnedders replied:
As mentioned in DMs: note `remote_to_name` v. `name_to_remote`; it's a different dict.

JonWBedard committed:
[git-webkit] Support pushing commits to user-specified remote
https://bugs.webkit.org/show_bug.cgi?id=284149
rdar://141025958

Reviewed by Sam Sneddon.

* Tools/Scripts/hooks/pre-push: Allow pushing to unregistered remotes, so long as we treat
those remotes as public.

Canonical link: https://commits.webkit.org/287491@main
